public/app/features/plugins/plugin_loader.ts

```diff
@@ -98,7 +99,17 @@ exposeToPlugin('jquery', jquery);
 exposeToPlugin('d3', d3);
 exposeToPlugin('rxjs', rxjs);
 exposeToPlugin('rxjs/operators', rxjsOperators);
-exposeToPlugin('react-router-dom', reactRouter);
+// This is the trick.
+// ==================
+// Plugins that still use "react-router-dom@v5" don't depend on react-router directly, so they will not use this import.
+// (The react-router-dom@v5 that we expose for them depends on the core react-router package internally.)
+//
+// Plugins that would like to use "react-router-dom@v6" will need to bundle "react-router-dom",
+// however they cannot bundle react-router - that would mean we have two instances of react-router
+// in the app, which would break it. As "react-router-dom-v5-compat" re-exports everything from "react-router-dom@v6",
+// which re-exports everything from "react-router@v6", we can expose a compatible v6 version of the router to plugins.
+exposeToPlugin('react-router', reactRouterCompat);
+exposeToPlugin('react-router-dom', reactRouterDom);
```

Comment:
Yeah exactly, they will need to bundle (not externalise) "react-router-dom", otherwise they will always receive v5.

```typescript
// If it was like this, then all the current plugins which externalise "react-router-dom" would receive the v6 versions of <Route> for example, or wouldn't have access to the <Switch> component. The reason for this is that the compat package re-exports basically everything from v6, plus adds the v5 compatible <CompatRouter> and <CompatRoute> components. 
exposeToPlugin('react-router-dom', reactRouterCompat);

// Like this we keep exposing v5 to make sure that plugins which don't update their webpack config (the ones that start migrating to v6) would continue working.
exposeToPlugin('react-router-dom', reactRouterDom);
```